[MM-49921] Cloud export compliance

nickmisasi opened the pull request:
#### Summary
Adds a new generic AddressForm component that powers entering a shipping address in the payment modal. Upon receiving a compliance failed code of 422 from the backend while upgrading plans, a purchase pending modal will be displayed. Subsequent visits to the purchase modal will display that same pending screen, until the compliance screen has been manually acted upon on our side.

![image](https://user-images.githubusercontent.com/12176405/215595754-39b88eb4-adfd-49b8-8680-84d3e8185022.png)
![image](https://user-images.githubusercontent.com/12176405/215595786-326d2fad-7cd9-495e-93b3-9e49d3cf372e.png)
![image](https://user-images.githubusercontent.com/12176405/215595901-83f345d0-90ce-44a7-8fb2-00aa7984e677.png)



#### Ticket Link
https://mattermost.atlassian.net/browse/MM-49921

#### Related Pull Requests
<!--
List all PRs related to resolving a ticket. For instance, if you submitted a PR to `mattermost/mattermost-server`, please include it here.
-->
- Has server changes (please link here)
- Has mobile changes (please link here)

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs.

For an easier comparison of UI changes a table (template below) can be used.

|  before  |  after  |
|----|----|
| <insert before screenshot here> | <insert after screenshot here> |

-->

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note
none
```

mattermost-build commented:
E2E tests not automatically triggered, because PR has no approval yet. Please ask a developer to review and then try again to attach the QA label.

nickmisasi commented:
/update-branch

johndavidlugtu commented:
Good stuff @nickmisasi -- let's re-use the copy from here: https://mattermost.atlassian.net/browse/MM-49496?focusedCommentId=140976 to keep things consistent

neallred commented on components/purchase_modal/purchase_modal.tsx:
```diff
@@ -1072,6 +1166,57 @@ class PurchaseModal extends React.PureComponent<Props, State> {
+
+        if (this.props.isComplianceBlocked) {
+            return (
+                <Elements
```
Nit: I think we can avoid wrapping this in Stripe's `<Elements>` provider for this branch. It will avoid some unnecessary work in the case that they come back to this later while still in screening. I think the loadStripe block up above can also be skipped if they are in screening.

neallred commented on components/payment_form/address_form.tsx:
```diff
@@ -0,0 +1,160 @@
+
+        const newStateValue = {
+            [name]: value,
+        } as unknown as Pick<Address, keyof Address>;
```
This requires a dom attribute matching the internal type name, which is tricky to ensure it is right at first and easy to accidentally break in subsequent changes. Looks like there's a mismatch between `line` vs `address` 1 & 2 at the moment. Could we make this more explicit with something like the following?

```tsx
    const handleInputChange = (key: keyof Address) => (
        event:
        | React.ChangeEvent<HTMLInputElement>
        | React.ChangeEvent<HTMLSelectElement>,
    ) => {
        const newStateValue = {
           [key]: value,
        }
        ...
     }
  
  ...
  <Input
                    name='address'
                    type='text'
                    value={address.line1}
                    onChange={handleInputChange('line1')}
```

neallred commented on components/purchase_modal/purchase_modal.tsx:
```diff
@@ -1060,10 +1101,63 @@ class PurchaseModal extends React.PureComponent<Props, State> {
+                            // Setup the initial country based on their billing country, or USA.
+                            initialAddress={
+                                {
+                                    country:
```
Could we make a helper to create a full blank address? There's some issues with the type coercion:
* the underlying inputs that receive an `undefined` at first will change from being "uncontrolled" to "controlled" inputs as they get filled in. This goes against what React wants and creates console errors.
* It can be hard to ensure all the places that trust and assume the fields match the types won't explode or have unexpected behavior. AFAICT it works out at the moment, but its another place that lends itself to hard to detect regressions.

nickmisasi replied:
Added a helper. I always forget Typescript isn't like Go where it will create an entire struct including all fields with zero values

neallred commented on components/payment_form/address_form.tsx:
```diff
@@ -0,0 +1,160 @@
+        setAddress({...address, ...newStateValue});
+
+        const {onAddressChange} = props;
+        if (onAddressChange) {
```
Nit: We can skip the onAddressChange null check since this is a required prop, and this component can't do anything meaningful without it.

neallred commented on components/payment_form/address_form.tsx:
```diff
@@ -0,0 +1,160 @@
+
+const AddressForm = (props: AddressFormProps) => {
+    const {formatMessage} = useIntl();
+    const [address, setAddress] = useState<Address>(props.initialAddress);
```
I've seen this pattern elsewhere, especially, in CWS, of taking in initial state from the parent, making a local copy, maintaining the copy, and sending changes out to the parent. It seems indirect and hard to tell what the source of truth is to me.

This works as is so its not a blocking change, but I think it would be simpler and easier to understand to pass in the parent state and always read it, and rely on the parent handler to update the state that will be sent back in. We always send in part of the parent state already in the form of the partial `{country: ...}` object. Always reading the parent would help us be more meticulous in the parent about sending in a fully fleshed out object instead of a type coerced object, and would let us avoid the `| null` in the parent to begin with.

nickmisasi replied:
I've updated the whole component to make it controlled. Did it the way it was initially because I basically just pulled the form out of the `payment_form` and adjusted it

neallred reviewed (changes requested):
LGTM overall! The blocking things to me are:
* Having a smaller svg image for the screening flow
* looks like with the `subscribeCloudSubscription` parameters changes, some of the existing usages need to be updated to fix type check errors
* I think we want to say screening will be finished in 3 days instead of 1.

nickmisasi commented on components/common/svg_images_components/compliance_screen_failed_svg.tsx:
```diff
@@ -0,0 +1,8686 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
I grabbed this from the Figma designs for this same flow on the portal side. I too thought it was way bigger than it should be but forgot to go back to it. Will take a look

M-ZubairAhmed commented on components/purchase_modal/process_payment_setup.tsx:
```diff
@@ -312,6 +333,33 @@ class ProcessPaymentSetup extends React.PureComponent<Props, State> {
+        case ProcessState.FAILED_COMPLIANCE_SCREEN:
+            return (
+                <IconMessage
+                    title={t(
```
wondering if `t` should be used here instead of this.props.intl ?

neallred replied:
The `IconMessage` component supports two apis. The API being used asks the user provide only an id, and the IconMessage component renders it as a FormattedMessage (without providing a defaultMessage). This is how a lot of the existing usage is.

There is also a formatted<PropName> api, that requests the user pass in an already internationalized ReactNode. In general I think the component should move towards removing all usages of the id-only API because of the pitfall of not providing a defaultMessage: languages not providing the translation string will make errors.

That said, there are other existing usages of the id-only api so I'm 0/5 on whether we want to make this usage use the formatted* API in this PR.

M-ZubairAhmed replied:
Yup no worries at all just wanted to double check its usage

annelieseklein commented:
Looks great!

1. @nickmisasi here's an example in terms of size of illustration and width of body copy text https://www.figma.com/file/Aob5ffHtfjKzQoebR93ycA/Cloud-Freemium?node-id=508%3A436984&t=MqtksV0Hnsp90QMH-1
2. the illustration you have seems to have her arm cut out - can we fix that? let me know if you need a new illustration
3. see here for how the checkbox item should look like in terms of sizing and font: https://www.figma.com/file/HU7OqEjUyvfqe5vOY0Ice7/Untitled?node-id=0%3A35228&t=ZSfhNAzuocPpZJ1k-1

nickmisasi commented on i18n/en.json:
```diff
@@ -303,6 +303,9 @@
   "admin.billing.subscription.cloudTrialBadge.daysLeftOnTrial": "{daysLeftOnTrial} trial days left",
   "admin.billing.subscription.cloudYearlyBadge": "Yearly",
+  "admin.billing.subscription.complianceScreenFailed.button": "Continue with free version",
+  "admin.billing.subscription.complianceScreenFailed.subtitle": "We are processing your cloud upgrade and will get back to you within 1 business day. In the meantime, please continue to enjoy the free version.",
```
Wrote this back in August, when it _was_ 1 business day 😂 fixed

nickmisasi commented:
@annelieseklein 

I've set a max-width of 560px on the IconMessage component. Note that this will take effect for all occurrences of this component like the upgrade success/failure screen, downgrade success/failure screen, etc.

![image](https://user-images.githubusercontent.com/12176405/216172545-44b3b21d-526d-4ef9-9abc-59459fa984ab.png)


I used this illustration https://www.figma.com/file/GtAkFsG8unkQkcIudmAJwE/Export-Compliance?node-id=65%3A1346&t=NDQmUeVCjX0LMWqe-4 which is what we currently have in the portal. If you'd like something else just shoot me a link 

I'll get the checkbox styling cleaned up and get back to you

nickmisasi commented:
@neallred I updated the subscribeCloudSubscription to optionally take the shipping address since downgrades/monthly won't actually need to provide this information. 

Going to wait on Anneliese before looking into the SVG size 

Otherwise I think I've addressed everything

neallred reviewed (approved):
LGTM pending resolution of the SVG question. Thanks, @nickmisasi !

marianunez commented on components/purchase_modal/index.ts:
```diff
@@ -48,6 +49,7 @@ function mapStateToProps(state: GlobalState) {
         invoices: getCloudDelinquentInvoices(state),
         isCloudDelinquencyGreaterThan90Days: isCloudDelinquencyGreaterThan90Days(state),
         isFreeTrial: subscription?.is_free_trial === 'true',
+        isComplianceBlocked: subscription?.compliance_blocked === 'true',
```
Is there a related CWS PR that includes this new field?

nickmisasi replied:
I thought I had requested review on them but I guess I did not. Should be there now.

johndavidlugtu commented on i18n/en.json:
```diff
@@ -303,6 +303,10 @@
   "admin.billing.subscription.cloudTrialBadge.daysLeftOnTrial": "{daysLeftOnTrial} trial days left",
   "admin.billing.subscription.cloudYearlyBadge": "Yearly",
+  "admin.billing.subscription.complianceScreenFailed.button": "Continue with free version",
+  "admin.billing.subscription.complianceScreenFailed.subtitle": "We are processing your cloud upgrade and will get back to you within 3 business days. In the meantime, please continue to enjoy the free version.",
```
> Good stuff @nickmisasi -- let's re-use the copy from here: https://mattermost.atlassian.net/browse/MM-49496?focusedCommentId=140976 to keep things consistent

Bringing this here so it's in-context @nickmisasi

nickmisasi replied:
`Your transaction is being reviewed`

`We will check things on our side and get back to you within 3 days once your Cloud Subscription upgrade is approved. In the meantime, please feel free to continue using the free version of our product.`

What you linked mentions licenses. Does `Cloud Subscription upgrade` in place of `license` work?

Is the CTA ok?

johndavidlugtu replied:
* Cloud subscription upgrade

CTA: Continue with Cloud Free

nickmisasi replied:
@johndavidlugtu all set
![image](https://user-images.githubusercontent.com/12176405/216411061-d67c4dc0-2c1f-48de-9748-3bcc60ebd682.png)

johndavidlugtu reviewed (approved):
Awesome, thanks
